### ISSUE TITLE
iter_with_drain false positive when no ownership of the vector

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#8538]
### Summary

The `iter_with_drain` lint says to replace `vec.drain(..)` with `vec.into_iter()`. It sounds like what this has in mind is the `impl<T> IntoIterator for Vec<T>`, but only the owner of the vec could call that, which does not work in general for all uses of `drain(..)`.

Replacing `vec.drain(..)` with `vec.into_iter()` when vec is a mutable reference ends up calling `impl<T> IntoIterator for &mut Vec<T>` which has a different element type than `drain(..)`, so that usually wouldn't work. Also, clippy would immediately trigger `into_iter_on_ref` in that case, telling you to remove the `.into_iter()` in favor of `.iter_mut()`.

### Lint Name

iter_with_drain

### Reproducer

```rust
pub struct Struct(Vec<String>);

pub fn repro(vec: &mut Vec<String>) -> Struct {
    Struct(vec
        .drain(..)
        .filter(|_i| /* ... */ true)
        .collect())
}
```

```console
warning: `drain(..)` used on a `Vec`
 --> src/main.rs:5:10
  |
5 |         .drain(..)
  |          ^^^^^^^^^ help: try this: `into_iter()`
  |
  = note: `#[warn(clippy::iter_with_drain)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#iter_with_drain
```

Following the suggestion causes an error.

```console
error[E0277]: a value of type `Vec<String>` cannot be built from an iterator over elements of type `&mut String`
    --> src/main.rs:4:12
     |
4    |       Struct(vec
     |  _____------_^
     | |     |
     | |     required by a bound introduced by this call
5    | |         .into_iter()
6    | |         .filter(|_i| /* ... */ true)
7    | |         .collect())
     | |__________________^ value of type `Vec<String>` cannot be built from `std::iter::Iterator<Item=&mut String>`
```

### Version

```text
rustc 1.61.0-nightly (285fa7ecd 2022-03-14)
binary: rustc
commit-hash: 285fa7ecd05dcbfdaf2faaf20400f5f92b39b3c6
commit-date: 2022-03-14
host: x86_64-unknown-linux-gnu
release: 1.61.0-nightly
LLVM version: 14.0.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by dtolnay** [comment]
Mentioning @ldm0 @flip1995 @giraffate since the lint is from https://github.com/rust-lang/rust-clippy/pull/8483.

**Comment by MingweiSamuel** [comment]
This also applies even if you have ownership of the `Vec`, but still need to keep it after draining. For example
```rust
fn example(mut vec: Vec<u32>) -> Vec<u32> {
   vec.drain(..).for_each(|x| println!("{}", x);
   vec
}

**Comment by Bromeon** [comment]
Can confirm, we experienced in godot-rust [with a `&mut Vec` parameter](https://github.com/godot-rust/godot-rust/blob/f60f7fbc49d94c2664d7598e11241e492fecc11b/gdnative-core/src/core_types/pool_array.rs#L136-L149) (the error appeared in a [CI run](https://github.com/godot-rust/godot-rust/runs/5559983020?check_suite_focus=true) of clippy nightly).

**Comment by ViliamVadocz** [comment]
This issue should be reopened because clippy is still giving false positives.

Reproducible example:
```rust
fn main() {
    let mut my_vec = Vec::new();
    for i in 0..10 {
        my_vec.push(i);
        for x in my_vec.drain(..) {
            println!("{x}");
        }
    }
}
```

False positive:
```
warning: `drain(..)` used on a `Vec`
   --> bug/src/main.rs:5:25
    |
  5 |         for x in my_vec.drain(..) {
    |                         ^^^^^^^^^ help: try: `into_iter()`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#iter_with_drain
```

Clippy version: `clippy 0.1.82 (2cbbe8b8 2024-07-28)`

**Comment by Jarcho** [comment]
That's a distinctly different issue due to conflicting borrows. A new issue should be opened if one already isn't.

**Comment by Avi-D-coder** [comment]
It's still happening.
